File: php/deduplicator.php
<?php
  /**
   * Load JSON file from given path
   * 
   * @param string $file_path
   * @return array|null JSON data as an associative array
   */
  function load_json($file_path): mixed{
    try {
      if (!file_exists(filename: $file_path)) {
        throw new Exception(message: "File not found");
      }

      $json_data = file_get_contents(filename: $file_path);
      $data = json_decode($json_data, associative: true);

      if (json_last_error() !== JSON_ERROR_NONE) {
        throw new Exception(message: "Invalid JSON data");
      }

      return $data;
    } catch (Exception $e) {
      echo "Error: " . $e->getMessage() . PHP_EOL;
      return null;
    }
  }

  /**
   * Format and ecord log of changes made to the leads
   * 
   * 
   * @param array $original
   * @param array $updated
   * @return array Log of changes made to the leads 
   */
  function log_changes($original, $updated): array {
    $changes = [];
    foreach ($updated as $key => $newValue) {
      $oldValue = $original[$key] ?? null;
      if ($oldValue !== $newValue) {
        $changes[] = [
          "field" => $key,
          "from" => $oldValue,
          "to" => $newValue
        ];
      }
    }
    return $changes;
  }


  /**
   * Deduplicate an array of onjects by a given key
   * 
   * @param string $key Key to deduplicate by
   * @param array $data Array of objects to deduplicate
   * @param array $log Log of changes made to the leads
   * @return array Deduplicated array of objects
   */
  function deduplicate_by_key(string $key, array $data, array &$log): array {
    $deduplicated = [];

    foreach ($data as $index => $record) {
      $value = $record[$key];

      if (!isset($deduplicated[$value])) {
        $deduplicated[$value] = ['record' => $record, 'index' => $index];
      } else {
        $current = $deduplicated[$value];
        $currentRecord = $current['record'];
        $currentIndex = $current['index'];

        if (
          $record['entryDate'] > $currentRecord['entryDate'] ||
          ($record['entryDate'] === $currentRecord['entryDate'] && $index > $currentIndex)
        ) {
          $deduplicated[$value] = ['record' => $record, 'index' => $index];

          $log[] = [
          "source" => $currentRecord,
          "updated" => $record,
          "changes" => log_changes(original: $currentRecord, updated: $record),
          ];
        }
      }
    }

    $deduplicated_leads = [];
    foreach ($deduplicated as $updated_record) {
      $deduplicated_leads[] = $updated_record['record'];
    }

    return $deduplicated_leads;
  }

  function deduplicate($data, $keys=['email', '_id']): array {
    $changeLog = [];
    foreach ($keys as $key) {
      $data = deduplicate_by_key(key: $key, data: $data, log: $changeLog);
    }
    return ['leads' => $data, 'changeLog' => $changeLog];
  }


  function main(): void {
    $options = getopt(short_options: "", long_options: ["input_file:", "output_file:"]);
    
    if (!isset($options['input_file']) || !isset($options['output_file'])) {
        echo "Usage: php script.php --input_file=<input_file> --output_file=<output_file>\n";
        exit(1);
    }

    $inputFile = $options['input_file'];
    $outputFile = $options['output_file'];

    $data = load_json(file_path: $inputFile);
    $leads = $data['leads'];

    $result = deduplicate(data: $leads);
    $deduplicatedData = $result['leads'];
    $changeLog = $result['changeLog'];

    file_put_contents(filename: $outputFile, data: json_encode(value: ["leads" => $deduplicatedData, "change_log" => $changeLog], flags: JSON_PRETTY_PRINT));

    echo "Deduplicated data saved to $outputFile\n";
}

  main();
?>

File: python/deduplicator.py
import json
import argparse

def load_json(file_path):
    '''
    Load JSON file from the given path
    args:
        file_path: str: path to JSON file
    return: 
        dict: JSON data
    '''
    try:
        with open(file_path, 'r') as file:
            return json.load(file)
    except Exception as e:
        print(f"Error loading JSON file: {e}")
        return None
    


def log_changes(original, updated):
    '''
    Log changes between two dictionaries
    args:
        original: dictionary before being updated
        updated: dictionary after being updated
    return:
        list: list of changes
    '''
    changes = []
    for key in updated.keys():
        if original[key] != updated[key]:
            changes.append({
                "field": key,
                "from": original[key],
                "to": updated[key]
            })
    return changes


def deduplicate_by_key(key, data, log):
    '''
    Deduplicate an array of dictionaries by a given key
    args:
        key: str: key to deduplicate by
        data: list: array of dictionaries
        log: list: log of changes
    return:
        list: deduplicated array of dictionaries
    '''
    deduplicated = {} # store deduplicated records
    for index, record in enumerate(data): # use enumerate to keep track of index
        value = record[key]
        if value not in deduplicated:
            deduplicated[value] = (record, index)
        else:
            current_record, current_index = deduplicated[value]
            # compare entryDates and indices to keep the most recent record
            if (record['entryDate'] > current_record['entryDate'] 
                or (record['entryDate'] == current_record['entryDate'] 
                    and index > current_index)):
                deduplicated[value] = (record, index)
                # log changes
                log.append({
                    "source": current_record,
                    "updated": record,
                    "changes": log_changes(current_record, record)
                })
    return [item[0] for item in deduplicated.values()]


def deduplicate(data, keys=['email', '_id']):
    '''
    Deduplicate JSON records
    
    I chose to deduplicate by 1 key at a time to keep the code simple and easy to understand. Another approach would be to deduplicate by email and _id at the same time, but that would require more complex array logic and nested loops to be able to do comparisons and matching the objects correctly by both ids.
    
    This approach is more understandable and easier to debug. If the data set is very large, this approach may not be the best one. It also allows for deduplicating by any key we desire, as the keys are passed as arguments.
    
    args:
        data: list: array of dictionaries
        keys: list: keys to deduplicate by
    return:
        list: deduplicated array of dictionaries
        list: log of changes
    '''
    changeLog = []
    for key in keys:
        data = deduplicate_by_key(key, data, changeLog)
    return data, changeLog


def main():
    parser = argparse.ArgumentParser(description="Deduplicate JSON records.")
    parser.add_argument("input_file", help="Path to input JSON file")
    parser.add_argument("output_file", help="Path to output JSON file")
    parser.add_argument("--log", help="Path to log file", default="deduplication_log.json")
    args = parser.parse_args()

    data = load_json(args.input_file)
    leads = data['leads']

    deduplicated_data, change_log = deduplicate(leads)

    with open(args.output_file, 'w') as output:
        json.dump({"leads": deduplicated_data}, output, indent=4)

    with open(args.log, 'w') as log_file:
        json.dump(change_log, log_file, indent=4)

    print(f"Deduplicated data saved to {args.output_file}")
    print(f"Log saved to {args.log}")


if __name__ == '__main__':
    main()
